x-pack/legacy/plugins/uptime/server/lib/adapters/monitor_states/elasticsearch_monitor_states_adapter.ts

```diff
@@ -29,11 +29,121 @@ interface LegacyMonitorStatesQueryResult {
   afterKey: any | null;
 }
 
+interface LegacyMonitorStatesRecentCheckGroupsQueryResult {
+  checkGroups: string[];
+  afterKey: any | null;
+}
+
+const legacyStatesQuerySize = 50;
+
 export class ElasticsearchMonitorStatesAdapter implements UMMonitorStatesAdapter {
   constructor(private readonly database: DatabaseAdapter) {
     this.database = database;
   }
 
+  // This query returns the most recent check groups for a given
+  // monitor ID.
+  private async runLegacyMonitorStatesRecentCheckGroupsQuery(
+    request: any,
+    query: any,
+    searchAfter?: any
+  ): Promise<LegacyMonitorStatesRecentCheckGroupsQueryResult> {
+    const body = {
+      query: {
+        bool: {
+          filter: [
+            {
+              // We check for summary.up to ensure that the check group
+              // is complete. Summary fields are only present on
+              // completed check groups.
+              exists: {
+                field: 'summary.up',
+              },
+            },
+            query,
+          ],
+        },
+      },
+      sort: [
+        {
+          '@timestamp': 'desc',
+        },
+      ],
+      size: 0,
+      aggs: {
+        monitors: {
+          composite: {
+            size: legacyStatesQuerySize,
+            sources: [
+              {
+                monitor_id: {
+                  terms: {
+                    field: 'monitor.id',
+                  },
+                },
+              },
+            ],
+          },
+          aggs: {
+            top: {
+              top_hits: {
+                sort: [
+                  {
+                    '@timestamp': 'desc',
+                  },
+                ],
+                _source: {
+                  includes: ['monitor.check_group', '@timestamp', 'agent.id'],
+                },
+                // The idea here is that we want to get enough documents to get all
+                // possible agent IDs. Doing that in a deterministic way is hard,
+                // but all agent IDs should be represented in the top 50 results in most cases.
+                // There's an edge case here where a user has accidentally configured
+                // two agents to run on different schedules, but that's an issue on the user side.
+                size: 50,
```

Comment:
WDYT about extracting this to a constant value as well?

Reply:
I'm not sure if it's worth extracting for a single value, I think it makes the code more confusing. Esp. since this isn't a somewhat arbitrary number like page size. It's chosen due to it's suitability for the expected cardinality of the locations.

Reply:
++ that is fine, I figured you'd view it that way. This is ok by me as well.